SortingBot/Assets/Scripts/CoroutineActions/Action.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright 2021-2022 The SeedV Lab.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using UnityEngine;
+
+namespace CoroutineActions {
+  // An action contains one or more Unity coroutine tasks that need to be run together.
+  public class Action {
+    private readonly List<ITask> _tasks = new List<ITask>();
+    private readonly List<Coroutine> _runningCoroutines = new List<Coroutine>();
+
+    // The action is done if and only if all its task coroutines have finished.
+    public bool IsRunning => _runningCoroutines.Any(coroutine => !(coroutine is null));
+
+    private MonoBehaviour _hostObject;
+    private Mutex _mutex = new Mutex();
+
+    // hostObject must be a Unity MonoBehaviour object that can start a coroutine. tasks may contain
+    // one or more coroutine tasks.
+    public Action(MonoBehaviour hostObject, IEnumerable<ITask> tasks) {
+      Debug.Assert(!(hostObject is null) && !(tasks is null));
+      _hostObject = hostObject;
+      foreach (var task in tasks) {
+        if (!(task is null)) {
+          _tasks.Add(task);
+          _runningCoroutines.Add(null);
+        }
+      }
+      Debug.Assert(_tasks.Count > 0);
+    }
+
+    // Tries to start all the task coroutines in a thread-safe way. Returns true if all the task
+    // coroutines have been started.
+    public bool TryStart() {
+      _mutex.WaitOne();
```

Comment:
Is this mutex needed? Will `TryStart` and `Stop` methods be called from different threads?

Reply:
Good suggestion! The mutex logic was inherited from some old experiments. For now, we don't need it at all, since all actions are started and stopped by the queue.